Feature/improve security

vegetable-bits opened the pull request.

vegetable-bits committed:
Remove jquery dependency

vegetable-bits committed:
Merge pull request #1 from vegetable-bits/feature/remove-jquery

Remove jquery dependency

vegetable-bits committed:
Upgrade Project
- enable es6 modules
- add vdcode config
- add composer.json

vegetable-bits committed:
Optimize frontend scripts
- Use ES6 modules
- Split javascript and load it with block.json

vegetable-bits committed:
Merge pull request #2 from vegetable-bits/feature/es6-frontend-scripts

Feature/es6 frontend scripts

vegetable-bits committed:
Optimize CSS, no !important anymore

vegetable-bits committed:
Merge pull request #3 from vegetable-bits/feature/es6-frontend-scripts

Optimize CSS, no !important anymore

vegetable-bits committed:
Fix critical authorization bypass in user settings

- Add proper authorization check to prevent users from editing other users' profiles
- Add nonce verification for enhanced security
- Only administrators can edit other users' profiles
- Remove insecure CORS header from user settings endpoint

Security: Fixes privilege escalation vulnerability

vegetable-bits committed:
Remove insecure CORS headers from AJAX endpoints

- Remove 'Access-Control-Allow-Origin: *' headers from all AJAX handlers
- Eliminates security risk of unrestricted cross-origin requests
- Affects registration, password reset request, and password reset handlers

Security: Fixes potential CORS misconfiguration vulnerability

vegetable-bits committed:
Remove dangerous extract() function usage

- Remove extract() call in return_view() method to prevent variable overwriting
- View files already use safe direct array access (['key'])
- Eliminates potential code injection and variable collision risks

Security: Fixes variable overwriting vulnerability

vegetable-bits committed:
Improve weak activation code generation

- Replace predictable sha1(email + time()) with wp_generate_password()
- Use 32-character cryptographically secure random activation codes
- Eliminates predictable patterns and timing attacks
- No special characters to avoid URL encoding issues

Security: Fixes weak activation token generation

vegetable-bits committed:
Add comprehensive password strength validation

- Add validate_password_strength() method to Helper class
- Enforce minimum 8 characters with uppercase, lowercase, numbers, and special chars
- Block common weak passwords (password, 123456, qwerty, etc.)
- Apply validation to user registration and password updates
- Provide clear error messages for password requirements

Security: Prevents weak password usage across the plugin

vegetable-bits committed:
Enhance rate limiting and IP detection

- Improve IP detection to handle proxies, load balancers, and Cloudflare
- Add rate limiting for password reset requests (3 attempts per hour)
- Use proper IP validation with private/reserved range exclusion
- Support X-Forwarded-For, CF-Connecting-IP, and other proxy headers
- Fallback handling for environments where IP detection fails

Security: Prevents brute force attacks via proxy bypassing and password reset abuse

vegetable-bits committed:
Enhance session security and logout handling

- Add secure logout cleanup with session destruction and transient clearing
- Implement open redirect prevention in logout URLs
- Add session fixation protection with ID regeneration after login
- Set secure session cookie parameters (HttpOnly, Secure, SameSite)
- Track last login IP and timestamp for security monitoring
- Clear login attempt counters on successful login/logout
- Validate redirect URLs to prevent open redirect vulnerabilities

Security: Comprehensive session management and logout security improvements

vegetable-bits committed:
Add comprehensive input validation and security checks

- Add enhanced email validation with suspicious pattern detection
- Add comprehensive username validation with reserved name blocking
- Validate email domains and prevent malformed addresses
- Block usernames with consecutive special characters
- Prevent XSS patterns in email addresses
- Apply enhanced validation to registration and user settings
- Add length limits and character restrictions for usernames
- Block system-reserved usernames (admin, root, api, etc.)

Security: Prevents malicious input and improves data integrity

vegetable-bits commented:
Hi @Arkenon,

here are the security issues i've found. One of then may allow unprivileged users to takeover admins. I'm sorry, just realizing it includes my previous merged too. So you have to cherry pick then. It's time for a rest :-)

Arkenon commented:
@vegetable-bits Thanks for your support. I'm going on vacation for a week. I won't have access to my laptop for the week. I'll check the pull requests when I get back.

Arkenon commented:
@vegetable-bits  Thanks again for your contribution! I've merged your pull request after reviewing the changes.

To keep the codebase consistent, I will remove the backslashes (\) before WordPress function calls. Both styles work, but consistency helps maintain readability.

Additionally, these changes make the plugin more secure by implementing better coding practices and following WordPress security standards.

Note: I noticed that Plugin Check is showing warnings due to ini_set() usage. I'll be replacing this with the native session_start() function in the next update:

`if ( ! headers_sent() ) {
    // WordPress session start function
    if ( ! session_id() ) {
        session_start([
            'cookie_httponly' => true,
            'cookie_secure' => is_ssl(),
            'cookie_samesite' => 'Strict'
        ]);
    }
}`

Arkenon commented:
I will publish a new version soon. Feel free to share your WordPress.org username so I can add you as a contributor.

vegetable-bits commented:
Hi @Arkenon,

there was some more places you have to change, for example:
`https://github.com/Arkenon/frontend-login-and-registration-blocks/blob/master/composer.json#L2`

It still points to my github repository. I've added this line to be able install the plugin direcly from my github with composer. You should change it to your github. There was also some dupplicated code i still wanted to optimize, not sure you saw it?

Anyway, thanks for merging!
Btw. I've converted the API already to the newer WP-JSON API instead admin-ajax.php calls, but i've stoped work on it for now and its not finished yet.

Arkenon commented:
Hi, I will review all changes again before publishing the new version.
